File: lib/spec/damageModel.js
function valueOf(source, keys, fallback = "") {
  if (!source) {
    return fallback;
  }

  for (const key of keys) {
    if (source[key] !== undefined && source[key] !== null && source[key] !== "") {
      return source[key];
    }
  }

  return fallback;
}

function toNumber(value, fallback = 0) {
  const number = Number(value);

  return Number.isFinite(number) ? number : fallback;
}

export function toRatio(percent) {
  return percent / 100;
}

export function toPercent(ratio) {
  return ratio * 100;
}

export function roundPercent(value) {
  if (!Number.isFinite(value)) {
    return 0;
  }

  return Math.round((value + Number.EPSILON) * 100) / 100;
}

export function criticalRateLimitFromStats(criticalStats) {
  const limit = valueOf(criticalStats, ["CriticalRateLimit", "criticalRateLimit"], null);

  if (!limit || !valueOf(limit, ["IsActive", "isActive"], false)) {
    return {
      isActive: false,
      capPercent: 100,
      overflowConversionRatePercent: 0,
      maxConvertedEvolutionDamagePercent: 0
    };
  }

  return {
    isActive: true,
    sourceName: valueOf(limit, ["SourceName", "sourceName"], ""),
    capPercent: toNumber(valueOf(limit, ["CapPercent", "capPercent"], 100), 100),
    overflowConversionRatePercent: toNumber(valueOf(limit, ["OverflowConversionRatePercent", "overflowConversionRatePercent"], 0), 0),
    maxConvertedEvolutionDamagePercent: toNumber(valueOf(limit, ["MaxConvertedEvolutionDamagePercent", "maxConvertedEvolutionDamagePercent"], 0), 0)
  };
}

export function effectiveCriticalRatePercent(critRatePercent, criticalRateLimit) {
  return Math.max(0, Math.min(critRatePercent, criticalRateLimit.capPercent));
}

export function convertedEvolutionDamagePercent(critRatePercent, criticalRateLimit) {
  if (!criticalRateLimit.isActive) {
    return 0;
  }

  const overflowPercent = Math.max(0, critRatePercent - criticalRateLimit.capPercent);
  const convertedPercent = overflowPercent * criticalRateLimit.overflowConversionRatePercent / 100;

  return Math.min(convertedPercent, criticalRateLimit.maxConvertedEvolutionDamagePercent);
}

export function totalEvolutionDamagePercent({ fixedEvolutionDamagePercent = 0, critRatePercent = 0, criticalRateLimit }) {
  return fixedEvolutionDamagePercent + convertedEvolutionDamagePercent(critRatePercent, criticalRateLimit);
}

export function evolutionDamageMultiplier({
  fixedEvolutionDamagePercent = 0,
  currentCritRatePercent = 0,
  baseCritRatePercent = 0,
  criticalRateLimit
}) {
  const currentEvolutionDamage = totalEvolutionDamagePercent({
    fixedEvolutionDamagePercent,
    critRatePercent: currentCritRatePercent,
    criticalRateLimit
  });
  const baseEvolutionDamage = totalEvolutionDamagePercent({
    fixedEvolutionDamagePercent,
    critRatePercent: baseCritRatePercent,
    criticalRateLimit
  });

  return (1 + toRatio(currentEvolutionDamage)) / (1 + toRatio(baseEvolutionDamage));
}

export function criticalAverageMultiplier({
  critRatePercent = 0,
  critDamageBonusPercent = 0,
  criticalOutgoingDamagePercent = 0,
  criticalRateLimit
}) {
  const effectiveCritRate = toRatio(effectiveCriticalRatePercent(critRatePercent, criticalRateLimit));
  const criticalHitMultiplier = (2 + toRatio(Math.max(0, critDamageBonusPercent)))
    * (1 + toRatio(Math.max(0, criticalOutgoingDamagePercent)));

  return 1 + effectiveCritRate * (criticalHitMultiplier - 1);
}

File: backend/src/main/java/com/rosaeng/sangdamso/spec/DamageModel.java
package com.rosaeng.sangdamso.spec;

import tools.jackson.databind.JsonNode;

public final class DamageModel {

    private DamageModel() {
    }

    public static double toRatio(double percent) {
        return percent / 100.0;
    }

    public static double toPercent(double ratio) {
        return ratio * 100.0;
    }

    public static double roundPercent(double value) {
        if (!Double.isFinite(value)) {
            return 0;
        }

        return Math.round((value + Math.ulp(1.0)) * 100.0) / 100.0;
    }

    public static CriticalRateLimit criticalRateLimitFromStats(JsonNode criticalStats) {
        JsonNode limit = child(criticalStats, "CriticalRateLimit", "criticalRateLimit");

        if (limit == null || limit.isNull() || !booleanValue(limit, false, "IsActive", "isActive")) {
            return new CriticalRateLimit(false, "", 100, 0, 0);
        }

        return new CriticalRateLimit(
            true,
            textValue(limit, "", "SourceName", "sourceName"),
            numberValue(limit, 100, "CapPercent", "capPercent"),
            numberValue(limit, 0, "OverflowConversionRatePercent", "overflowConversionRatePercent"),
            numberValue(limit, 0, "MaxConvertedEvolutionDamagePercent", "maxConvertedEvolutionDamagePercent")
        );
    }

    public static double effectiveCriticalRatePercent(double critRatePercent, CriticalRateLimit criticalRateLimit) {
        return Math.max(0, Math.min(critRatePercent, criticalRateLimit.capPercent()));
    }

    public static double convertedEvolutionDamagePercent(double critRatePercent, CriticalRateLimit criticalRateLimit) {
        if (!criticalRateLimit.isActive()) {
            return 0;
        }

        double overflowPercent = Math.max(0, critRatePercent - criticalRateLimit.capPercent());
        double convertedPercent = overflowPercent * criticalRateLimit.overflowConversionRatePercent() / 100.0;

        return Math.min(convertedPercent, criticalRateLimit.maxConvertedEvolutionDamagePercent());
    }

    public static double totalEvolutionDamagePercent(
        double fixedEvolutionDamagePercent,
        double critRatePercent,
        CriticalRateLimit criticalRateLimit
    ) {
        return fixedEvolutionDamagePercent + convertedEvolutionDamagePercent(critRatePercent, criticalRateLimit);
    }

    public static double evolutionDamageMultiplier(
        double fixedEvolutionDamagePercent,
        double currentCritRatePercent,
        double baseCritRatePercent,
        CriticalRateLimit criticalRateLimit
    ) {
        double currentEvolutionDamage = totalEvolutionDamagePercent(
            fixedEvolutionDamagePercent,
            currentCritRatePercent,
            criticalRateLimit
        );
        double baseEvolutionDamage = totalEvolutionDamagePercent(
            fixedEvolutionDamagePercent,
            baseCritRatePercent,
            criticalRateLimit
        );

        return (1 + toRatio(currentEvolutionDamage)) / (1 + toRatio(baseEvolutionDamage));
    }

    public static double criticalAverageMultiplier(
        double critRatePercent,
        double critDamageBonusPercent,
        double criticalOutgoingDamagePercent,
        CriticalRateLimit criticalRateLimit
    ) {
        double effectiveCritRate = toRatio(effectiveCriticalRatePercent(critRatePercent, criticalRateLimit));
        double criticalHitMultiplier = (2 + toRatio(Math.max(0, critDamageBonusPercent)))
            * (1 + toRatio(Math.max(0, criticalOutgoingDamagePercent)));

        return 1 + effectiveCritRate * (criticalHitMultiplier - 1);
    }

    private static JsonNode child(JsonNode node, String... keys) {
        if (node == null || node.isNull()) {
            return null;
        }

        for (String key : keys) {
            JsonNode value = node.get(key);

            if (value != null && !value.isNull() && !(value.isTextual() && value.asString().isEmpty())) {
                return value;
            }
        }

        return null;
    }

    private static String textValue(JsonNode node, String fallback, String... keys) {
        JsonNode value = child(node, keys);
        return value == null ? fallback : value.asString();
    }

    private static double numberValue(JsonNode node, double fallback, String... keys) {
        JsonNode value = child(node, keys);

        if (value == null || value.isNull()) {
            return fallback;
        }

        if (value.isNumber()) {
            return value.asDouble();
        }

        try {
            return Double.parseDouble(value.asString().replace(",", "").trim());
        } catch (NumberFormatException exception) {
            return fallback;
        }
    }

    private static boolean booleanValue(JsonNode node, boolean fallback, String... keys) {
        JsonNode value = child(node, keys);

        if (value == null || value.isNull()) {
            return fallback;
        }

        if (value.isBoolean()) {
            return value.asBoolean();
        }

        return Boolean.parseBoolean(value.asString());
    }

    public record CriticalRateLimit(
        boolean isActive,
        String sourceName,
        double capPercent,
        double overflowConversionRatePercent,
        double maxConvertedEvolutionDamagePercent
    ) {
    }
}
